### PR TITLE
Reproducibility: improve the check and add recorder-time validation

### PR DESCRIPTION
**Pull request by Ladicek**
See commit messages for more details.

The first commit actually makes a lot of tests fail early on when running in reproducibility checking mode. One of them was SmallRye Fault Tolerance, which made me wonder why the reproducibility check didn't fail before. This led me to discover that certain sources of non-determinism were hidden in the reproducibility checks by all Quarkus builds running in the same JVM. This provoked the second commit. It probably slows down the reproducibility check considerably, but is more faithful, so I'm on the fence here.

I didn't check how many failures the second commit uncovers (maybe it's the same set as the first commit?), but with the third commit, the SmallRye Fault Tolerance extension passes :-)

CC @reaver585

### CODE REVIEW

**Comment by manovotn** [comment]
> This led me to discover that certain sources of non-determinism were hidden in the reproducibility checks by all Quarkus builds running in the same JVM. This provoked the second commit. It probably slows down the reproducibility check considerably, but is more faithful, so I'm on the fence here.

This is a really interesting observation. Did you check, roughly, how much slower that is?
I definitely think it would be worth taking a hit in execution time to get that coverage but having at least an estimation would be neat.

> The first commit actually makes a lot of tests fail early on when running in reproducibility checking mode.

:100: 
I didn't realize we'd even be able to check it. This is simply brilliant!

**Comment by mkouba** [comment]
> The first commit actually makes a lot of tests fail early on when running in reproducibility checking mode.

+100

> This provoked the second commit. It probably slows down the reproducibility check considerably, but is more faithful, so I'm on the fence here.

I think that no matter how much slower it is, we'll need to get it in, even if hidden behind a config switch (if the slowdown is too drastic and impractical).

**Comment by reaver585** [comment]
Thanks a lot @Ladicek. This is awesome. @gsmet also mentioned in the past that we could try to enforce the "no `HashMap`s or `HashSet`s" rule via the bytecode recorder, and this goes a step beyond that with the `hashCode` override.
And, I was also worried about all the edge cases that weren't caught by the deployments tests we have, so this will for sure help with that.

I'm polishing the CI pull request in the meantime, so we could also check how much longer the tests would take with the new approach you have here, spinning up a separate JVM. Perhaps it's not that bad 😄 

+1 from my side.

File: test-framework/junit-internal/src/main/java/io/quarkus/test/ReproducibilityTestLauncher.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import org.junit.platform.launcher.core.LauncherFactory;
+import org.junit.platform.launcher.listeners.SummaryGeneratingListener;
+
+/**
+ * Minimal JUnit launcher for running a single test class in a forked JVM.
+ * Used by the reproducibility check in {@link AbstractQuarkusExtensionTest}.
+ */
+class ReproducibilityTestLauncher {
+    public static void main(String[] args) {
+        var request = LauncherDiscoveryRequestBuilder.request()
```
Comment (manovotn) [inline test-framework/junit-internal/src/main/java/io/quarkus/test/ReproducibilityTestLauncher.java:16]: Personally, I wouldn't mind a little less `var` in this PR but I understand that might be just my opinion :)

**Review by manovotn — DISMISSED** [review]
I am not a fan of `var` but otherwise it is looking good 👍 

If you are too busy, I am happy to try the test time diff with this change, just let me know.
Either way, I think we want to get it in, we should just be aware of how big the impact is.

**Comment by Ladicek** [comment]
> > This led me to discover that certain sources of non-determinism were hidden in the reproducibility checks by all Quarkus builds running in the same JVM. This provoked the second commit. It probably slows down the reproducibility check considerably, but is more faithful, so I'm on the fence here.
> 
> This is a really interesting observation. Did you check, roughly, how much slower that is? I definitely think it would be worth taking a hit in execution time to get that coverage but having at least an estimation would be neat.

On a fresh build of Quarkus, I ran the following command in the `deployment` module of the SmallRye Fault Tolerance extension: `mvn clean test -Dquarkus-internal.test.reproducibility-check -Dno-build-cache`. The results are:

| branch | time |
| --- | --- |
| `main` | ~50 seconds |
| `improve-reproducibility-test` | ~10 minutes |

An order of magnitude worse, that's pretty bad.

I then realized the previous test used a cached application model for all runs, which might have been there for a reason. Maybe that accounts for a big chunk of the time? I know we serialize / deserialize the app model elsewhere, so I made a similar change: before we start running the checks, curate the application, get the model from there, serialize it, and in each forked JVM, deserialize the model and pass it to the curation process. Turns out that didn't made too big of a difference, the tests execute in ~9 minutes. So this is not the way.

I do think this is valuable, but for now, I think we'll have to have 2 options:

1. run "fast" reproducibility checks, including what I'm adding here in the 1st commit
2. run "slow" but more faithful reproducibility checks -- I hope that after "fast" checks succeed, the "slow" checks will usually succeed as well or the number of changes required won't be too high

I'll refactor the 2nd commit to allow both options.

**Comment by mkouba** [comment]
+1 for "fast" and "slow" checks.

**Comment by Ladicek** [comment]
Rebased and added another system property: `quarkus-internal.test.reproducibility-check.cross-jvm`. If not set, the reproducibility check runs as before (the "fast" mode). If set, the reproducibility check runs each Quarkus build inside a new JVM (the "slow" mode).

File: test-framework/junit-internal/src/main/java/io/quarkus/test/ReproducibilityTestLauncher.java (hunk 1 above)
Comment (Ladicek) [reply test-framework/junit-internal/src/main/java/io/quarkus/test/ReproducibilityTestLauncher.java:16]: This is frankly a purely AI-generated class which I only had to fix one bug in and didn't touch otherwise. It's the least important part of this PR :-)

**Comment by gsmet** [comment]
@Ladicek it's 10 minutes just for the SRFT deployment tests? If so, yeah, it won't fly, there's no way we will be able to run the full CI in this mode, unfortunately. I wonder if we should profile it quickly to try to understand what's so slow?

@reaver585 I think it's yet another reason to use the approach I used for the quickstarts for the ITs. It fully builds a new app with a new JVM each time.

**Comment by Ladicek** [comment]
> @Ladicek it's 10 minutes just for the SRFT deployment tests? If so, yeah, it won't fly, there's no way we will be able to run the full CI in this mode, unfortunately. I wonder if we should profile it quickly to try to understand what's so slow?

Correct. I have no idea why it's so slow. I wasn't even able to debug the forked JVM properly (though I think the command was OK, I was able to connect to it with a debugger, but the breakpoints I set just didn't hit), so I'm not sure where would I start with profiling :-)

We could possibly get rid of the 2nd commit from this PR -- the 1st adds enough value on its own IMHO. I can try to figure out why it's so slow and maybe submit it as another PR later on.

**Comment by gsmet** [comment]
Given it's optional, I don't think it hurts to have it there and it can be useful if other people want to give it a try.

**Review by gsmet — APPROVED** [review]
Let's get this in.

@reaver585 there's a good chance this won't help you :).

**Comment by Ladicek** [comment]
> Let's get this in.
> 
> @reaver585 there's a good chance this won't help you :).

I think there's a good chance this _will_ help, due to the 1st commit. This is a super cheap check that uncovered a lot of bugs when I ran it.

Actually not a _lot_ -- it just found bugs in RESTEasy Reactive, so half of the remaining modules were simply skipped :-) Not sure how many of them are affected. Maybe not too many.
